controllers/release/release_adapter.go

```diff
@@ -0,0 +1,301 @@
+// the Get operation failed, an error will be returned.
+func (a *Adapter) getReleaseLink() (*v1alpha1.ReleaseLink, error) {
+	releaseLink := &v1alpha1.ReleaseLink{}
+	err := a.client.Get(a.context, types.NamespacedName{
+		Namespace: a.release.Namespace,
+		Name:      a.release.Spec.ReleaseLink,
+	}, releaseLink)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return releaseLink, nil
+}
+
+// getReleasePipelineRun returns the PipelineRun referenced by the Release being processed or nil if it's not found.
+// In the case the List operation fails, an error will be returned.
+func (a *Adapter) getReleasePipelineRun() (*v1beta1.PipelineRun, error) {
+	pipelineRuns := &v1beta1.PipelineRunList{}
+	opts := []client.ListOption{
+		client.MatchingLabels{
+			tekton.ReleaseNameLabel:      a.release.Name,
+			tekton.ReleaseWorkspaceLabel: a.release.Namespace,
+		},
+	}
+
+	err := a.client.List(a.context, pipelineRuns, opts...)
+	if err != nil {
+		return nil, err
+	} else if len(pipelineRuns.Items) == 0 {
+		return nil, nil
+	}
+
+	return &pipelineRuns.Items[0], nil
+}
+
+// getReleaseStrategy returns the ReleaseStrategy referenced by the given ReleaseLink. If the ReleaseStrategy is
+// not found or the Get operation fails, an error will be returned.
+func (a *Adapter) getReleaseStrategy(releaseLink *v1alpha1.ReleaseLink) (*v1alpha1.ReleaseStrategy, error) {
+	releaseStrategy := &v1alpha1.ReleaseStrategy{}
+	err := a.client.Get(a.context, types.NamespacedName{
+		Name:      releaseLink.Spec.ReleaseStrategy,
+		Namespace: releaseLink.Namespace,
+	}, releaseStrategy)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return releaseStrategy, nil
+}
+
+// getTargetReleaseLink returns the ReleaseLink targeted by ReleaseLink in the Release being processed. If a matching
+// ReleaseLink is not found or the List operation fails, an error will be returned.
+func (a *Adapter) getTargetReleaseLink() (*v1alpha1.ReleaseLink, error) {
+	releaseLink, err := a.getReleaseLink()
+	if err != nil {
+		return nil, err
+	}
+
+	releaseLinks := &v1alpha1.ReleaseLinkList{}
+	opts := []client.ListOption{
+		client.InNamespace(releaseLink.Spec.Target),
+		client.MatchingFields{"spec.target": releaseLink.Namespace},
+	}
+
+	err = a.client.List(a.context, releaseLinks, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, foundReleaseLink := range releaseLinks.Items {
+		if foundReleaseLink.Spec.Application == releaseLink.Spec.Application {
+			return &foundReleaseLink, nil
+		}
+	}
+
+	return nil, fmt.Errorf("no ReleaseLink found in target workspace '%s' with target '%s' and application '%s'",
+		releaseLink.Spec.Target, releaseLink.Namespace, releaseLink.Spec.Application)
+}
+
+// registerReleasePipelineRunStatus updates the status of the Release being processed by monitoring the status of the
+// associated release PipelineRun and marking those states in the Release. If the PipelineRun hasn't started/succeeded,
```

Comment:
`s/marking those states/marking its status/`? Aren't we just pulling the status of the pipeline run and adding that to the release? Isn't it always just one?

Reply:
Changed to `and setting the appropriate state in the Release`.